exec/java-exec/src/main/java/org/apache/drill/exec/work/foreman/Foreman.java

```diff
@@ -1221,6 +1223,12 @@ public void interrupted(final InterruptedException e) {
      *   to the user
      */
     public void moveToState(final QueryState newState, final Exception ex) {
+      // if the current thread is the foreman thread, throw an exception
+      // otherwise the foreman will be blocked forever on acceptExternalEvents
+      if (myThreadRef == Thread.currentThread()) {
```

Comment:
Well, Foreman does not call moveToState in common path but when it fails then it could. My point is instead of hacking the method to throw an exception if the thread is foreman we should release the latch and handle the exception in foreman#run. What I mean is

Foreman#run() { 
```
try {
  do sth
} catch(ex) {
  release latch 
  handleException(ex)
} finally {
  ...
}
```